File: example_usage.cpp
#include <iostream>
#include <string>
#include <chrono>
#include <thread>
#include <atomic>
#include <csignal>

#include "ximea_manager.h"

// Simple atomic to handle Ctrl+C
std::atomic<bool> keep_running(true);
void signal_handler(int) { keep_running = false; }

int main() {
    // 1. Setup signal handling for clean exit
    std::signal(SIGINT, signal_handler);

    // 2. Instantiate the Manager
    // This class coordinates the camera (1000Hz), recorder (1000Hz), and previewer (60Hz)
    XimeaManager manager;

    // 3. Initialize Settings
    // Params: save_fps, display_fps, codec ("h264", "h265", "av1")
    int save_fps = 1000;
    int display_fps = 60;
    std::string codec = "h265"; // Using HEVC for better high-speed compression

    std::cout << "--- Initializing XimeaManager ---" << std::endl;
    if (!manager.Initialize(save_fps, display_fps, codec)) {
        std::cerr << "Failed to initialize. Ensure XIMEA camera is connected and Above 4G Decoding is enabled." << std::endl;
        return 1;
    }

    // 4. Start Acquisition
    // This launches background threads for acquisition and encoding.
    // It returns immediately (asynchronous).
    // An optional error callback can be passed to handle runtime failures.
    auto error_callback = [](const std::string& msg) {
        std::cerr << "\n[CALLBACK ERROR] " << msg << std::endl;
        keep_running = false;
    };

    std::cout << "--- Starting High-Speed Capture ---" << std::endl;
    if (!manager.Start(".", "cpp_example_output", error_callback)) {
        std::cerr << "Failed to start capture." << std::endl;
        return 1;
    }

    std::cout << "Capture running for 10 seconds. Press Ctrl+C to stop." << std::endl;

    // 5. Telemetry Loop
    // Periodically poll the manager for real-time performance statistics.
    auto start_time = std::chrono::steady_clock::now();
    while (keep_running && manager.IsRunning()) {
        auto now = std::chrono::steady_clock::now();
        if (std::chrono::duration_cast<std::chrono::seconds>(now - start_time).count() >= 10) {
            break;
        }

        XimeaTelemetry tel = manager.GetTelemetry();

        // Print a simple live status line
        std::cout << "\r[CAM] Captured: " << tel.cam_captured 
                  << " Dropped: " << tel.cam_dropped 
                  << " | [REC] Encoded: " << tel.rec_encoded 
                  << " Queue: " << tel.rec_queue 
                  << " | [DISP] " << tel.prev_displayed << std::flush;

        std::this_thread::sleep_for(std::chrono::milliseconds(100));
    }

    // 6. Graceful Stop
    // This stops threads, flushes GStreamer pipelines, and closes the camera.
    std::cout << "\n--- Stopping ---" << std::endl;
    manager.Stop();
    std::cout << "Capture finished. Output saved to cpp_example_output.mp4" << std::endl;

    return 0;
}

File: python/example_usage.py
import sys
import os
import time

# 1. Add the build directory to the path so Python can find 'ximea_py.so'
build_path = os.path.abspath("./ximea_highspeed_app/build")
if not os.path.exists(build_path):
    build_path = os.path.abspath("./build")
sys.path.append(build_path)

try:
    import ximea_py
except ImportError as e:
    print(f"Error: Could not import ximea_py. Make sure it is built in {build_path}")
    print(f"Details: {e}")
    sys.exit(1)

def error_handler(message):
    """Callback function triggered by the C++ core on fatal errors."""
    print(f"\n[PYTHON CALLBACK] Fatal Error Detected: {message}")

def run_example():
    # 2. Instantiate the Manager
    manager = ximea_py.XimeaManager()

    # 3. Initialize Settings
    save_fps = 1000
    display_fps = 60
    codec = "h264" 

    # 3b. Configure Camera
    cam_cfg = ximea_py.CameraConfig()
    cam_cfg.width = 4096
    cam_cfg.height = 996
    cam_cfg.exposure_us = 900
    cam_cfg.gain_db = -1.0 # -1 means max analog gain
    cam_cfg.offset_x = 0
    cam_cfg.offset_y = 0

    print(f"--- Initializing XIMEA Manager ---")
    print(f"Target: {save_fps} FPS | Preview: {display_fps} FPS | Codec: {codec}")
    print(f"Camera: {cam_cfg.width}x{cam_cfg.height}, Exposure: {cam_cfg.exposure_us}us, Gain: {cam_cfg.gain_db}dB")
    
    if not manager.initialize(save_fps, display_fps, codec, cam_cfg):
        print("Failed to initialize camera. Check permissions and connection.")
        return

    # 4. Start Acquisition
    output_dir = "."
    file_prefix = "python_highspeed_test"
    
    print(f"--- Starting Capture ---")
    if not manager.start(output_dir, file_prefix, error_handler):
        print("Failed to start capture loop.")
        return

    print("Capture running. Monitoring telemetry...")
    print("Press 'r' in the preview window to toggle recording.")
    print("Press 'q' in the preview window or Ctrl+C to quit.\n")

    try:
        while manager.is_running():
            # 5. Fetch Real-Time Telemetry
            tel = manager.get_telemetry()
            rec_status = "[RECORDING]" if manager.is_recording() else "[IDLE]"
            idx = manager.get_recording_index()
            
            # Print a live status line
            status = (
                f"\r{rec_status} #{idx} | [CAM] Cap: {tel.cam_captured} | Drop: {tel.cam_dropped} "
                f"| [REC] Enc: {tel.rec_encoded} | Q: {tel.rec_queue} "
                f"| [DISP] {tel.prev_displayed}"
            )
            print(f"{status: <100}", end="", flush=True)
            
            time.sleep(0.1) # Update at 10Hz
            
    except KeyboardInterrupt:
        print("\nInterrupted by user.")

    # 6. Graceful Stop
    print(f"\n\n--- Stopping Capture ---")
    manager.stop()
    print(f"Capture finished. Video saved to {file_prefix}.mkv")

if __name__ == "__main__":
    run_example()
